src/server/command/worldedit/blueprint.zig

```diff
@@ -0,0 +1,193 @@
+
+	if(source.commandData.clipboard) |clipboard| {
+		const storedBlueprint = clipboard.store(main.stackAllocator);
+		defer main.stackAllocator.free(storedBlueprint);
+
+		const fileName: []const u8 = ensureBlueprintExtension(main.stackAllocator, args.items[1]);
+		defer main.stackAllocator.free(fileName);
+
+		var cwd = std.fs.cwd();
+
+		_ = cwd.makeDir("blueprints") catch null;
+
+		var blueprintsDir = try cwd.openDir("blueprints", .{});
+		defer blueprintsDir.close();
+
+		std.log.info("Saving clipboard to blueprint file: {s}", .{fileName});
+		source.sendMessage("#00ff00Saving clipboard to blueprint file: {s}", .{fileName});
+
+		try blueprintsDir.writeFile(.{
+			.sub_path = fileName,
+			.data = storedBlueprint,
+			.flags = .{.lock = .exclusive},
+		});
+	} else {
+		source.sendMessage("#ff0000Error: No clipboard content to save.", .{});
+	}
+}
+
+fn ensureBlueprintExtension(allocator: NeverFailingAllocator, fileName: []const u8) []const u8 {
+	if(!std.ascii.endsWithIgnoreCase(fileName, ".blp")) {
+		return std.fmt.allocPrint(allocator.allocator, "{s}.blp", .{fileName}) catch unreachable;
+	} else {
+		return allocator.dupe(u8, fileName);
+	}
+}
+
+fn blueprintDelete(args: List([]const u8), source: *User) !void {
+	if(args.items.len < 2) {
+		source.sendMessage("#ff0000**/blueprint delete** requires FILENAME argument.", .{});
+		return;
+	}
+	if(args.items.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint delete**. Expected 1 argument, FILENAME.", .{});
+		return;
+	}
+
+	const fileName: []const u8 = ensureBlueprintExtension(main.stackAllocator, args.items[1]);
+	defer main.stackAllocator.free(fileName);
+
+	var cwd = std.fs.cwd();
+
+	_ = cwd.makeDir("blueprints") catch null;
+
+	var blueprintsDir = try cwd.openDir("blueprints", .{});
+	defer blueprintsDir.close();
+
+	try blueprintsDir.deleteFile(fileName);
+
+	std.log.info("Deleted blueprint file: {s}", .{fileName});
+	source.sendMessage("#ff0000Deleted blueprint file: {s}", .{fileName});
+}
+
+fn blueprintList(_: List([]const u8), source: *User) !void {
+	var cwd = std.fs.cwd();
+
+	_ = cwd.makeDir("blueprints") catch null;
+
+	var blueprintsDir = try cwd.openDir("blueprints", .{.iterate = true});
+	defer blueprintsDir.close();
+
+	var directoryIterator = blueprintsDir.iterate();
+	var index: i32 = 0;
+
+	while(try directoryIterator.next()) |entry| {
+		if(entry.kind != .file) break;
+		if(!std.ascii.endsWithIgnoreCase(entry.name, ".blp")) break;
+
+		source.sendMessage("#ffffff{}#00ff00 {s}", .{index, entry.name});
+		index += 1;
+	}
+}
+
+fn blueprintLoad(args: List([]const u8), source: *User) !void {
+	if(args.items.len < 2) {
+		source.sendMessage("#ff0000**/blueprint load** requires FILENAME argument.", .{});
+		return;
+	}
+	if(args.items.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint load**. Expected 1 argument, FILENAME.", .{});
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```

Comment:
The lock should be around only the necessary bits of data (→the `source.commandData.clipboard != null` check and writing the new blueprint to the clipboard).
This makes it harder to run into deadlocks (like what if `source.sendMessage` also needs to lock the mutex in the future?) and it's faster too.